File: Isogeny.c
//isogeny.h   このプログラムの名前
#include<stdio.h>
#include<stdlib.h>
#include"Efp2.c"
#include<gmp.h>

#define aa0 "0000F37AB34BA0CEAD94F43CDC50DE06AD19C67CE4928346E829CB92580DA84D7C36506A2516696BBE3AEB523AD7172A6D239513C5FD2516 000196CA2ED06A657E90A73543F3902C208F410895B49CF84CD89BE9ED6E4EE7E8DF90B05F3FDB8BDFE489D1B3558E987013F9806036C5AC"
#define ba0 "0001CD28597256D4FFE7E002E87870752A8F8A64A1CC78B5A2122074783F51B4FDE90E89C48ED91A8F4A0CCBACBFA7F51A89CE518A52B76C 000147073290D78DD0CC8420B1188187D1A49DBFA24F26AAD46B2D9BB547DBB6F63A760ECB0C2B20BE52FB77BD2776C3D14BCBC404736AE4"
#define ka 11
#define kb 2
#define eA "000000D8"
#define eB "00000089"
#define la 2
#define lb 3

Fp2 aA0;
Fp2 bA0;

int ea, eb;

void Isogeny_set_a0(void){

    Fp2_set_str(&Aa0, aa0);
    Fp2_set_str(&Ba0, ba0);
 
}

void Isogeny_set_e(void){
    ea = 
    Fp_set_str(&ea, eA);
    Fp_set_str(&eb, eB);
}

void Isogeny_gets(ec2 *S, ec2 *P, ec2 *Q, int k, Fp2 *ap){
    // SaやSbを求めます
    //Sa = Pa + ka*Qa
    ec2 U; 

    Fp2_mgecSCM(&U, Q, k, ap);
    Fp2_mgecA(S, P, &U, ap);

    free(&U);
}

void isogeny_changea(Fp2 *nexta, Fp2 *alpha){
    Fp2 one, two;
    Fp2_set_ui(&one, "1 0");
    Fp2_set_ui(&two, "2 0");
    Fp2 temp;
    Fp2_to_Mont(alpha, alpha);

    Fp2_mul(&temp, alpha, alpha);
    Fp2_mul(&temp, &temp, &two);
    Fp2_sub(&temp, &one, &temp);
    Fp2_mul(nexta, &two, &temp);

    free(&one);
    free(&two);
    free(&temp);
    Fp2_from_Mont(nexta);
}

void Isogeny_changeb(Fp2 *nexta, Fp2 *beta, Fp2 *olda){
    Fp2 six;
    Fp2_set_ui(&six, "6 0");

    Fp2 temp; 
    Fp2 temp2;

    Fp2_to_Mont(olda, olda);
    Fp2_to_Mont(beta, beta);

    Fp2_mul(&temp, olda, beta);
    Fp2_mul(&temp2, &six, beta);
    Fp2_mul(&temp2, &temp2, beta);
    Fp2_sub(&temp, &temp, &temp2);
    Fp2_add(&temp, &temp, &six);
    Fp2_mul(&temp, &temp, beta);

    Fp2_set(nexta, &temp);
    Fp2_from_Mont(nexta);
}

void Isogeny_nextp(ec2 *ans, ec2 *P, Fp2 *alpha, int l){
    //同種写像Φを求めます
    //P...点 alpha...スタート地点 l...AliceですかBobですか
    ec2 U; 

    Fp2 temp; 
    Fp2 bumbo; 
    Fp2 temp2; 

    Fp2 two; 
    Fp2_set_ui(&two, "2 0");

    Fp2_to_Mont(alpha, alpha);
    Fp2_to_Mont(&P->x, &P->x);
    Fp2_to_Mont(&P->y, &P->y);

    if(l == 2){
        //alice       
        //分子
        
        Fp2_mul(&temp, &P->x, &P->x);
        Fp2_mul(&temp, &temp, alpha);
        Fp2_sub(&temp, &temp, &P->x);
        

        //分母
        Fp2_sub(&bumbo, &P->x, alpha);
        
        Fp2_inv(&bumbo, &bumbo);

        
        //掛け算
        Fp2_mul(&U.x, &temp, &bumbo);

        
        //set
        Fp2_set(&ans->x,&U.x);
        Fp2_from_Mont(&ans->x);
        
    }else{
        //bob

        Fp three;
        Fp_set_str(&three, "3");
        
        Fp2 one; 
        Fp2_set_ui(&one, "1 0");

        //
        Fp2_mul(&temp, alpha, &P->x);
        Fp2_sub(&temp, &temp, &one);
        Fp2_mul(&temp, &temp, &temp);
        Fp2_mul(&temp, &temp, &P->x);
        

        //
        Fp2_sub(&bumbo, &P->x, alpha);
        Fp2_mul(&bumbo, &bumbo, &bumbo);
        Fp2_inv(&bumbo, &bumbo);
        

        //
        Fp2_mul(&U.x, &temp, &bumbo);

        //
        Fp2_set(&ans->x, &U.x);
        Fp2_from_Mont(&ans->x);

    }
}

void isogeny_mgec3(ec2 *R, ec2 *P, fp2 *ap){
    ec2 U; 
    
    Fp2 temp, temp2, temp3; 

    Fp four, three;
    Fp_set_str(&four, "4");
    Fp_set_str(&three, "3");

    Fp2 six, four2, three2, one;
    
    Fp2_set_str(&six, "6 0");
    Fp2_set_str(&four2, "4 0");
    Fp2_set_str(&three2, "3 0");
    Fp2_set_str(&one, "1 0");

    Fp2_to_Mont(&P->x, &P->x);
    Fp2_to_Mont(&P->y, &P->y);
    Fp2_to_Mont(ap, ap);

    Fp2 bunshi, bumbo;

    Fp2_mul(&temp, &P->x, &P->x);
    Fp2_mul(&temp, &temp, &P->x);
    Fp2_mul(&temp, &temp, &P->x);
    //x^4

    Fp2_mul(&temp2, &six, &P->x);
    Fp2_mul(&temp2, &temp2, &P->x);
    //6x^2

    
    Fp2_mul(&temp3, &P->x, &four2);
    Fp2_mul(&temp3, &temp3, &A);
    //4ax^3

    Fp2_sub(&temp, &temp, &temp2);
    Fp2_sub(&temp, &temp, &temp3);
    Fp2_sub(&temp, &temp, &three2);
    //()

    Fp2_mul(&temp, &temp, &temp);
    //()^2

    Fp2_mul(&temp, &temp, &P->x);
    //x*()^2

    Fp2_set(&bunshi, &temp);


    Fp2_mul(&temp, &P->x, &P->x);
    Fp2_mul(&temp, &temp, &P->x);
    Fp2_mul(&temp, &temp, &P->x);
    Fp2_mul(&temp, &temp, &three2);
    //3x^4

    Fp2_mul(&temp3, &P->x, &P->x);
    Fp2_mul(&temp3, &temp3, &P->x);
    Fp2_mul(&temp3, &temp3, &four2);
    Fp2_mul(&temp3, &temp3, &A);
    //4ax^3
    
    Fp2_add(&temp, &temp, &temp2);
    Fp2_add(&temp, &temp, &temp3);
    Fp2_sub(&temp, &temp, &one);
    //()

    Fp2_mul(&temp, &temp, &temp);
    //()^2

    Fp2_inv(&temp, &temp);
    // /

    Fp2_set(&bumbo, &temp);

    Fp2_mul(&U.x, &bunshi, &bumbo);
    
    Fp2_set(&R->x, &U.x);
    Fp2_from_Mont(&R->x);
}

File: isogeny.py
# isogeny.py
import define
import fp2

class efp2_t:
    def __init__(self, x, y, infinity):
        self.x = fp2.fp2_t(x0, x1)
        self.y = fp2.fp2_t(x0, x1)
        self.infinity = 0
    def info(self):
        print("_________________________")
        print(f'x0:{self.x0}')
        print(f'x1:{self.x1}')
# def print_fp(x: fp_t):
#     print("x0:{x.x0}")
    def __repr__(self):
        print(self.x0)
        print(self.x1)
    def __str__(self):
        return str(self.x0) + "_" + str(self.x1)

# aの更新
def isogeny_changea(oldb: fp2.fp2_t, alpha: fp2.fp2_t):
    one = fp2.fp2_t(1, 0)
    two = fp2.fp2_t(2, 0)

    tmp = fp2.fp2_t(0, 0)

    #//a = 2(1 - 2alpha^2)
    
    tmp = alpha * alpha

    print(f'alpha ** 2 =  {tmp}')
    
    tmp = tmp * two

    print(f'tmp =  {tmp}')
    
    tmp = one - tmp

    print(f'tmp = {tmp}')
    
    nexta = two * tmp

    #//b = oldb * alpha

    #// ここの処理の有無でA. Costello B. SIDHspec 楕円曲線式を切り替える

    nextb = oldb * alpha

    return nexta.mod(), nextb.mod()

# 2-isogeny
def isogeny_nextp(ans: efp2_t, P: efp2_t, alpha:fp2_t, l:int):
    # //同種写像Φを求めます
    # //P...点 alpha...スタート点 l...AliceですかBobですか
    U,V = efp2_t(P.x.x0,P.x.x1,P.y.x0,P.y.x1,0, efp2_t(0,0,0,0,0))

    #//U.infinity =  P->infinity

    temp, temp2, bumbo = fp2.fp2_t(0,0),fp2.fp2_t(0,0), fp2.fp2_t(0,0)

    if l == 2:
        # //alice       
        # //分子

        # //xφ2(P) =xP**2 * x2 − xP/ xP − x2

        temp = U.x * U.x
        temp = temp * alpha
        temp = temp - U.x

        #//fp2_printf("bunshi = ", &temp)

        #//分母
        bumbo = P.x - alpha

        #//fp2_printf("bumbo = ", &bumbo);

        bumbo = bumbo.inv()

        #//fp2_printf("inv bumbo = ", &bumbo)

        #//掛け算
        V.x = temp * bumbo

        #//set
        ans.x = V.x

        #//y 分子 xP**2* alpha − 2*xP*alpha**2 + alpha
        temp = U.x * U.x
        temp = temp * alpha
        temp = temp - U.x

        temp2 = U.x * U.x
        temp2 = temp2 * alpha
        temp2 = temp2 * alpha

        temp = temp - temp2

        temp = temp + alpha

        #//fp2_printf("bunshi =", &temp);

        #//分母　(xP − alpha)**2
        bumbo = U.x - alpha
        bumbo = bumbo * bumbo

        #//fp2_printf("bumbo =", &bumbo);

        bumbo = bumbo.inv()

        #//fp2_printf("inv bumbo =", &bumbo);

        temp = temp * bumbo
        V.y = temp * U.y

        # // c**2 = alpha からcを求めて掛ける
        # //(C), (D)の切り替え
        if fp2_legendre(alpha) == 1:
            fp2_sqrt(&temp, alpha)
            V.y = V.y * temp

        ans.y = V.y

    # # }else{
    # #     //bob

    # #     mpz_t three;
    # #     mpz_init(three);
    # #     mpz_set_ui(three, 3);

    # #     //printf("bob\n");

    # #     fp2_t one, temp3, temp4; 
    # #     fp2_init(&one);
    # #     fp2_set_ui(&one, 1);
    # #     fp2_init(&temp3);
    # #     fp2_init(&temp4);

    # #     //
    # #     fp2_mul(&temp, alpha, &U.x);
    # #     fp2_sub(&temp, &temp, &one);
    # #     fp2_mul(&temp, &temp, &temp);
    # #     fp2_mul(&temp, &temp, &U.x);
    # #     //fp2_printff(tempp);

    # #     //
    # #     fp2_sub(&bumbo, &P->x, alpha);
    # #     fp2_mul(&bumbo, &bumbo, &bumbo);
    # #     fp2_inv(&bumbo, &bumbo);
    # #     //fp2_printff(bumbop);

    # #     //
    # #     fp2_mul(&V.x, &temp, &bumbo);

    # #     //
    # #     fp2_set(&ans->x, &V.x);

    # #     //y 分子 (xP*a − 1)(xP**2*a − 3*xP*a**3 + xP + a)
    # #     // printf("alpha:");
    # #     // fp2_printf_s(alpha);
    # #     // printf("U:");
    # #     // Efp2_PrintEC2_s(&U);

    # #     fp2_mul(&temp, &U.x, alpha);
    # #     fp2_sub(&temp, &temp, &one);

    # #     //////////////////////////////////////

    # #     fp2_mul(&temp2, &U.x, alpha);
    # #     fp2_mul(&temp2, &temp2, &U.x);

    # #     fp2_mul(&temp3, &U.x, alpha);
    # #     fp2_mul(&temp3, &temp3, alpha);
    # #     fp2_mul(&temp3, &temp3, alpha);

    # #     fp2_add(&temp4, &temp3, &temp3);
    # #     fp2_add(&temp4, &temp4, &temp3);

    # #     fp2_sub(&temp2, &temp2, &temp4);

    # #     fp2_add(&temp2, &temp2, &U.x);

    # #     fp2_add(&temp2, &temp2, alpha);

    # #     ///////////////////////////////////////

    # #     fp2_mul(&temp, &temp, &temp2);

    # #     //分母 (xP − x3)3
    # #     fp2_sub(&temp3, &U.x, alpha);
    # #     fp2_mul(&bumbo, &temp3, &temp3);
    # #     fp2_mul(&bumbo, &bumbo, &temp3);

    # #     fp2_inv(&bumbo, &bumbo);

    # #     ///////////////////////////////////////

    # #     fp2_mul(&V.y, &temp, &bumbo);

    # #     fp2_mul(&V.y, &V.y, &U.y);

    # #     fp2_set(&ans->y, &V.y);
